### DISCUSSION TITLE
Is there a way to go from Symbol->SyntaxNode-Symbol

### DISCUSSION THREAD

**Discussion by Practical-UVM-Step-By-Step** [MikePopoloski/slang#841] [General]
Mike: I saw the other answers in the discussion, but they didnt fit what I was asking. I am trying to write a check to detect #delays in forever blocks in task code.

Please consider this example code:
```package der_pkg;
import base_pkg::*;

	class myclass;
	string name;
	int der_int;
	string der_string;
	function new(string namec="a");
		name = namec;
		der_int = a;
		der_string = base_string;
		set_values(10,"string_param");
	endfunction
	task put;
		$display("hello there");
	endtask
	task mytask;
		int a;
		int b = 10;
		forever begin
			a = 20;
			wait(der_int ==1);
			put();
			#10;
		end
		
	endtask
	endclass
endpackage
```

I am using an additional check like this in slang::tidy
```

struct MainVisitor : public TidyVisitor, ASTVisitor<MainVisitor, true, false> {
    explicit MainVisitor(Diagnostics& diagnostics) : TidyVisitor(diagnostics) {}
      void handle(const ForeverLoopStatement& symbol){

         const slang::syntax::SyntaxNode* sym_syntax = symbol.syntax;
         detect(*sym_syntax,symbol);

      }
      MainVisitor& detect(const slang::syntax::SyntaxNode& node,const ForeverLoopStatement& symbol ) {
          size_t childCount = node.getChildCount();
          for (size_t i = 0; i < childCount; i++) {
              if (auto childNode = node.childNode(i); childNode) {
                  auto mykind = childNode->kind;
                  if (mykind == slang::syntax::SyntaxKind::DelayControl) {

                      diags.add(diag::ForeverBlocksDelay, childNode->getFirstToken().location());
                  }
                  detect(*childNode, symbol);
              }
          }
          return *this;
      }
};
```

MY question:
1) is there a better/clean way to do this? 
2) in detect(), I've effectively gone from Symbols to the Syntax Tree.  I passed in the Forever symbol because I would lose the opportunity to go back to the symbols data structure. 

I looked in Symbol.h and SyntaxNode.h. There didnt seem to be a way to go between the two. Maybe it was by design, and I havent understood why yet. 

Thank you.

**Comment by MikePopoloski** [comment]
You shouldn't have to go back to the syntax nodes. You can look for expressions that have TimingControls attached (delays are one type of timing control) and report an error if you find it. In general you can go back from symbols/expressions/statements (i.e. the AST) to the original syntax nodes but you can't go forward from the syntax nodes to the AST.
